### PR TITLE
Loading function for Anipose data

### PR DESCRIPTION
**Pull request by vigji**
## Description

Adding support to load adipose-triangulated data.

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
Currently, adipose-triangulated data cannot be loaded. I love the autocorrect here.

**What does this PR do?**
Implement simple loading function for anipose-generated data. Anipose saves a csv file with x, y, z, score (confidence) plus more info to a `csv` file. Currently, this basic info is parsed. Additional info that is stored in the file, which could be loaded in principle but would not fit any of `movement` boxes atm:
 - number of cameras each key point was triangulated from. This is a useful piece of data; if there's space for arbitrary arrays in the dataset I would add it
 - back projection error: another important metrics that would be useful to pack together.
 - info on the coordinate transformations implemented during triangulation, less important imho - but could also be dumped in. Not sure if we would end up to bloat the DataSet too much

## References
#124 

## How has this PR been tested?
So far just very coarsely tested on some anipose data I am generating. I am no proficient anipose user, and I'm happy to collect example data and feedback from anyone. Will try to dig into example data from the anipose repo to make sure everything can be loaded from there.

## Is this a breaking change?
No

## Does this PR require an update to the documentation?
Maybe once done

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [ ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Comment by adamltyson** [comment]
> I love the autocorrect here.

Thanks for adding support for fat data.

**Comment by niksirbi** [comment]
Thanks for opening this PR @vigji, I will take a look at it next week, and also think about the possibility of storing the additional data as arrays in the same dataset.

Off the top of my head we might take a 2-step-approach:
- in this PR, just load x, y, z + confidence (all of which are supported by current data structures).
- add the extra arrays in a future PR, after discussing how to represent those variables

Also, following this PR, and before it's released, we'd have to go through our functions (filtering, kinematics, vector utils) and make sure they don't freak out with 3D data. Many of them will be fine, but some of them have not been designed/tested with 3D in mind. But this is a bullet we have to bite anyway sooner or later, and sooner would be much easier.

**Comment by vigji** [comment]
Makes perfect sense! I will deal with anipose loading + test kinematic stuff on 3D model here, and leave additional datasets loading for another PR.

I could dig out some open source anipose data from the Anipose publication, what is your policy there? Should I just copy it to the GIN storage, ask first, or?

**Comment by niksirbi** [comment]
> I could dig out some open source anipose data from the Anipose publication, what is your policy there? Should I just copy it to the GIN storage, ask first, or?

With out current infrastructure, the easiest would be to just upload the data to GIN, if the license of the original data allows that. Push access to our GIN repo is restricted, but you should already have the required permission.

Long term, we are also considering to add [support for downloading publicly available datasets](https://github.com/neuroinformatics-unit/movement/issues/340), as long as they have persistent identifiers and reliable URLs, but this is not implemented yet. When we have that, it may no longer be necessary to mirror everything on GIN.

**Comment by vigji** [comment]
Dedicated validator and testing should have been implemented, let me know what you think and if you think docs should also be updated!

Kinematics remain to be tested

File: tests/test_unit/test_validators/test_files_validators.py
Comment (niksirbi) [inline tests/test_unit/test_validators/test_files_validators.py:219]: In this case it's `ValueError` for all cases, but you probably intented to keep this general, hence the parametrised `error_type`.

```suggestion
    with pytest.raises(error_type) as excinfo:
```

**Review by niksirbi — CHANGES_REQUESTED** [review]
Thanks a lot, @vigji!

I've done a first pass and left several comments—some of them quite nitpicky (and therefore optional).

## Public `load_poses` API

As you review my comments, you'll notice that many pertain to function names, parameter names, and docstrings in `load_poses.py`. One of `movement`'s core goals is to provide a unified interface for multiple tracking software packages, so I’ve aimed to enforce uniformity where it makes sense.

In addition to renaming certain elements, it would be great to find a way to incorporate "Anipose" as an option in the generic `load_poses.from_file()` function:

```python
def from_file(
    file_path: Path | str,
    source_software: Literal["DeepLabCut", "SLEAP", "LightningPose", "Anipose"],
    fps: float | None = None,
) -> xr.Dataset:
```

This function delegates the loading task to the appropriate software-specific function based on `source_software`. The challenge with "Anipose" lies in its additional keyword argument, `individual_name`. I’m unsure how best to integrate this into the generic `from_file` function. Here are a few potential approaches:

1. **Expose `individual_name` in `load_poses.from_file()`**, but make it clear in the docstring that this parameter only applies when `source_software` is "Anipose." While this solution is somewhat awkward, it preserves the functionality for other software sources.
2. **Convert it into a list—`individual_names`—that can also apply to other software sources.** This would allow users to override individual names in DLC/SLEAP files as well. However, this approach is more complex and would require significant changes to the implementations of other loaders. It could also be confusing, as we don’t currently allow similar overrides for keypoint names, which contradicts the logic of a "loader" function.
3. **Don’t expose `individual_name` in `load_poses.from_file()`.** Under this approach, users would always get the default "individual_0" for Anipose when using this function. If they need to specify a custom individual name during loading, they could use the software-specific `load_poses.from_anipose_file()` function. Alternatively, they could rename it later using xarray functionality.
4. **A variant of option 3: Allow custom kwargs to be passed through `load_poses.from_file()`.** These `**kwargs` would be fed directly into the underlying software-specific loader. I like this option because it adds flexibility, which might prove useful as we continue to add support for new software sources that may require additional parameters.

Let me know your thoughts on these options.

## Documentation
Once we've resolved the above issues, we'll need to update `docs/source/user_guide/input_output.md`. Specifically, we should add Anipose as a supported framework and include a corresponding tab in the "Loading pose tracks" section to demonstrate how the new functions should be used.

## Kinematics and 2D-Specific Functions
In response to your points:
> Kinematics remain to be tested.  
> From what I see in the kinematic functions:
> 
> - there should not be 2D-specific stuff in velocity computing;
> - there is a check for 2 spatial dimensions in the `compute_forward_vector`; how do we want to deal with this? Should we generalise the forward vector to 3D? Can that be done in a separate PR, and for now we just assume that for 3D data there is still a plane over which to project to compute those metrics? In my case, the mouse is still moving on a plane, so this is well defined. This might not be the case for swimming/flying/complex environments. But maybe this can be addressed separately, also with the contribution of people who deal with those kinds of data. I do not want to define the framework for data I do not work with!

I’ve reviewed all our functions, and the only ones that might present issues with 3D data are:

1. **`compute_forward_vector()` and its alias `compute_head_direction_vector()`**  
   These functions include a specific check for 2D data. I think this is fine for now, as they are clearly documented as 2D-specific. If a user passes 3D data, they’ll encounter an informative error. For now, we could open an issue to track the idea of "generalising forward vector to 3D." Generalisation might make sense, but we’d need to carefully consider its implications and how to implement it.

2. **`cart2pol` and `pol2cart` from `utils/vector.py`**  
   These functions conceptually only apply to 2D data (`x, y` ↔ `rho, phi`) and currently lack an explicit check for spatial dimensions. While we may eventually provide 3D equivalents (e.g., `cart2sph` and `sph2cart`), they could remain as separate functions. There's already been extensive discussion about this in [Zulip](https://neuroinformatics.zulipchat.com/#narrow/channel/406001-Movement/topic/coordinate.20systems.20and.20reference.20frames). For now, adding a check to restrict input spatial dimensions to 2 (similar to the one in `compute_forward_vector`) should suffice.

That said, it would be prudent to verify that all other kinematics, filtering, and vector utility functions behave correctly with 3D pose data. To achieve this, we would need to:

- Create a 3D variant of the `valid_poses_dataset_uniform_linear_motion` fixture and parameterise existing tests to include this fixture.
- Write additional 3D-specific tests where necessary.

Since this will require some effort, I suggest opening a dedicated issue to address these tasks separately after this PR is merged.

**Comment by vigji** [comment]
> Thanks a lot, @vigji!

Thank you super much for the very thorough PR review, I learn a lot from you people! And sorry for the sloppiness of some of the copypastes, I did it and made it to work and then forgot to clean up things

> I've done a first pass and left several comments—some of them quite nitpicky (and therefore optional).

While I implement those, on the high level stuff:

> In addition to renaming certain elements, it would be great to find a way to incorporate "Anipose" as an option in the generic `load_poses.from_file()` function:

Absolutely!


> 1. **Expose `individual_name` in `load_poses.from_file()`**, but make it clear in the docstring that this parameter only applies when `source_software` is "Anipose." While this solution is somewhat awkward, it preserves the functionality for other software sources.

quote awkward indeed though

> 2. **Convert it into a list—`individual_names`—that can also apply to other software sources.** This would allow users to override individual names in DLC/SLEAP files as well. However, this approach is more complex and would require significant changes to the implementations of other loaders. It could also be confusing, as we don’t currently allow similar overrides for keypoint names, which contradicts the logic of a "loader" function.

A bit of a mess

> 3. **Don’t expose `individual_name` in `load_poses.from_file()`.** Under this approach, users would always get the default "individual_0" for Anipose when using this function. If they need to specify a custom individual name during loading, they could use the software-specific `load_poses.from_anipose_file()` function. Alternatively, they could rename it later using xarray functionality.
> 4. **A variant of option 3: Allow custom kwargs to be passed through `load_poses.from_file()`.** These `**kwargs` would be fed directly into the underlying software-specific loader. I like this option because it adds flexibility, which might prove useful as we continue to add support for new software sources that may require additional parameters.

3 or 4. Default is good, but for me 4 is the best in terms of flexibility vs args completeness. I can only expect more idiosyncratic arguments as you support more and more files. One vote for  4, half vote for 3 :)

> 
> ## Documentation
> Once we've resolved the above issues, we'll need to update `docs/source/user_guide/input_output.md`. Specifically, we should add Anipose as a supported framework and include a corresponding tab in the "Loading pose tracks" section to demonstrate how the new functions should be used.

Ok!

> 1. **`compute_forward_vector()` and its alias `compute_head_direction_vector()`**
>    These functions include a specific check for 2D data. I think this is fine for now, as they are clearly documented as 2D-specific. If a user passes 3D data, they’ll encounter an informative error. For now, we could open an issue to track the idea of "generalising forward vector to 3D." Generalisation might make sense, but we’d need to carefully consider its implications and how to implement it.

Yes, I think it makes sense to postpone this

> 2. **`cart2pol` and `pol2cart` from `utils/vector.py`**
>    These functions conceptually only apply to 2D data (`x, y` ↔ `rho, phi`) and currently lack an explicit check for spatial dimensions. While we may eventually provide 3D equivalents (e.g., `cart2sph` and `sph2cart`), they could remain as separate functions. There's already been extensive discussion about this in [Zulip](https://neuroinformatics.zulipchat.com/#narrow/channel/406001-Movement/topic/coordinate.20systems.20and.20reference.20frames). For now, adding a check to restrict input spatial dimensions to 2 (similar to the one in `compute_forward_vector`) should suffice.

Ok

> That said, it would be prudent to verify that all other kinematics, filtering, and vector utility functions behave correctly with 3D pose data. To achieve this, we would need to:
> 
> * Create a 3D variant of the `valid_poses_dataset_uniform_linear_motion` fixture and parameterise existing tests to include this fixture.
> * Write additional 3D-specific tests where necessary.
> 
> Since this will require some effort, I suggest opening a dedicated issue to address these tasks separately after this PR is merged.

Ok, I will postpone this testing and adding the above caveat to the next PR then if you agree

**Comment by niksirbi** [comment]
> 3 or 4. Default is good, but for me 4 is the best in terms of flexibility vs args completeness. I can only expect more idiosyncratic arguments as you support more and more files. One vote for 4, half vote for 3 :)

Agreed re 4,  in fact I already need this flexibility (I need to add yet another kwarg for NWB files).

Sounds like we have a plan for this in general! The only thing I'd ask is to make sure you open issues for all the stuff we postpone.

File: tests/test_unit/test_validators/test_files_validators.py
Comment (vigji) [reply tests/test_unit/test_validators/test_files_validators.py:213]: done

**Comment by vigji** [comment]
ok, this should be good now!

File: movement/io/load_poses.py
Comment (niksirbi) [inline movement/io/load_poses.py:121]:
```suggestion
        Additional keyword arguments to pass to the software-specific
        loading functions that are listed under "See Also".
```

File: movement/io/load_poses.py
Comment (niksirbi) [inline movement/io/load_poses.py:796]:
```suggestion
```

`individual_name` appears twice in the docstring. I suggest deleting the first occurrence, so that the order of arguments in the docstring matches that of the function signature.

File: movement/io/load_poses.py
Comment (niksirbi) [inline movement/io/load_poses.py:794]: By convention we include the period `.` when mentioning file extensions in docstrings.
```suggestion
        Path to the Anipose triangulation .csv file
```

File: movement/io/load_poses.py
Comment (niksirbi) [inline movement/io/load_poses.py:744]: This is the old dimension order, please update to the current one.

 ```suggestion
     time, space, keypoints, individuals, and array of confidence scores
     with dimensions time, keypoints, individuals.

File: movement/io/load_poses.py
Comment (niksirbi) [inline movement/io/load_poses.py:742]: The word "confidence" doesn't appear in the actual column names, right? It's just keypoint1_score etc. I wonder if we should remove it from the columns we name here, to avoid the confusion of users expecting (or creating) dataframes with "confidence" among the column names. The fact that "keypoint1_score" refers to the confidence score should be obvious from the fact that we convert that into a confidence array (as we say in the following sentence).

File: movement/io/load_poses.py
Comment (niksirbi) [inline movement/io/load_poses.py:99]: Pleas also add "Anipose" to this parameter's description in the docstring.

**Review by niksirbi — APPROVED** [review]
Thanks for updating @vigji.

I've added a few comments as finishing touches.

This PR is still missing the updates on the docs input/output page.

File: movement/io/load_poses.py
Comment (vigji) [reply movement/io/load_poses.py:99]: done

**Comment by vigji** [comment]
Right, here we go! There should be everything now.

**Comment by niksirbi** [comment]
> Right, here we go! There should be everything now.

Very good, the anipose url had not been defined, but should be fine now, I've added it to `must_url_schemes` in Sphinx's `conf.py`. I will merge this once all CI checks have passed.

Thanks for doing all the work!

**Pull request merged**
